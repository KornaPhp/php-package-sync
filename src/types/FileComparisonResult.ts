/* eslint-disable no-unused-vars */

export enum ComparisonKind {
    PACKAGE_NOT_USED = 'missing_pkg',
    PACKAGE_VERSION_MISMATCH = 'pkg_version',
    PACKAGE_SCRIPT_NOT_FOUND = 'pkg_script',

    FILE_NOT_FOUND = 'missing_file',
    DIRECTORY_NOT_FOUND = 'missing_dir',

    FILE_DOES_NOT_MATCH = 'not_exact',
    FILE_NOT_SIMILAR_ENOUGH = 'not_similar',

    FILE_NOT_IN_SKELETON = 'extra_file',
    DIRECTORY_NOT_IN_SKELETON = 'extra_dir',

    ALLOWED_SIZE_DIFFERENCE_EXCEEDED = 'size_diff',
}

export interface FileComparisonResult {
    kind: ComparisonKind;
    score: string | number;
    name: string;
    context?: any;
    skeletonPath: string;
    repositoryPath: string;
}
