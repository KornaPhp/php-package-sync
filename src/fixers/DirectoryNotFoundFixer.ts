/* eslint-disable no-unused-vars */

import { mkdirSync } from 'fs';
import { ComparisonKind } from '../types/FileComparisonResult';
import Fixer from './Fixer';

export class DirectoryNotFoundFixer extends Fixer {
    public static handles = [ComparisonKind.DIRECTORY_NOT_FOUND];

    public description() {
        return 'creates a missing directory';
    }

    public fix(): boolean {
        if (!this.shouldPerformFix()) {
            return false;
        }

        mkdirSync(`${this.issue.repository.path}/${this.issue.name}`, { recursive: true });

        this.issue.resolve(this)
            .addResolvedNote(`created directory '${this.issue.name}'`);

        return true;
    }

    public static prettyName(): string {
        return 'create-dir';
    }
}
