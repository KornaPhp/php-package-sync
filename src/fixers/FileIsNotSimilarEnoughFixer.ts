/* eslint-disable no-unused-vars */

import { classOf } from '../lib/helpers';
import { Fixer } from './Fixer';

const readline = require('readline');
const util = require('util');

export class FileIsNotSimilarEnoughFixer extends Fixer {
    public static handles = []; //ComparisonKind.FILE_NOT_SIMILAR_ENOUGH];

    public description() {
        return 'file is out of sync with the skeleton version and user action is required.';
    }

    public async promptUser(fixers: string[]) {
        const rl = readline.createInterface({
            input: process.stdin,
            output: process.stdout,
        });

        const question = util.promisify(rl.question)
            .bind(rl);

        const askQuestion = async prompt => {
            let result;

            try {
                result = await question(`${prompt} `);
            } catch (err) {
                result = false;
            }

            return new Promise(resolve => resolve(result));
        };

        try {
            const input: string = <string>await askQuestion(`Run fixer '${fixers[0]}' on file '${this.issue.name}'? [y/N] `);

            if ((input.trim()
                .toLowerCase()[0] ?? 'n') !== 'y') {
                this.issue.resolve(classOf(this)
                    .prettyName())
                    .addResolvedNote('rejected running more fixers');

                this.issue.disableFixers();
                return;
            }

            this.issue.pending = false;

            this.issue.fixers[0].fix();
        } finally {
            rl.close();
            this.issue.pending = false;
        }
    }

    public fix(): boolean {
        if (!this.shouldPerformFix()) {
            return false;
        }

        this.issue.resolve(this);

        return false;
    }

    public async run() {
        this.issue.pending = true;
        this.issue.resolved = false;

        // const additionalFixers = this.issue.availableFixers.filter(name => name !== FileIsNotSimilarEnoughFixer.prettyName());

        // if (additionalFixers.length > 0) {
        //     await this.promptUser(additionalFixers);
        // }
    }

    public static prettyName(): string {
        return 'user-review';
    }
}
