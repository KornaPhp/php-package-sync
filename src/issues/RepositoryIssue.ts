/* eslint-disable no-unused-vars */

import { ComparisonKind } from '../types/FileComparisonResult';
import { Repository } from '../lib/Repository';
import { RepositoryFile } from '../lib/RepositoryFile';

export class RepositoryIssue {
    constructor(
        public result: any,
        public name: string,
        public srcFile: RepositoryFile | null,
        public destFile: RepositoryFile | null,
        public skeleton: Repository,
        public repository: Repository,
        public resolved: boolean = false,
        public note: string | null = null,
        public context: any | null = null,
    ) {
        //
    }

    get kind(): ComparisonKind {
        return this.result.kind;
    }

    get score(): number | string {
        return this.result.score;
    }
}
