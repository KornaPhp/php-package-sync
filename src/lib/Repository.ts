/* eslint-disable no-unused-vars */
/* eslint-disable @typescript-eslint/no-unused-vars */

import { basename } from 'path';
import { app } from '../Application';
//import { PackageIssue } from '../issues/PackageIssue';
import { RepositoryIssue } from '../issues/RepositoryIssue';
import { Composer } from './composer/Composer';
import { getFileList, isDirectory } from './helpers';
import { RepositoryFile } from './RepositoryFile';

export enum RepositoryKind {
    SKELETON = 'skeleton',
    PACKAGE = 'package',
}

export class Repository {
    protected composerData: Composer;
    protected loadedFiles = false;
    protected fileList: RepositoryFile[] = [];
    public issues: RepositoryIssue[] = [];

    constructor(public path: string, public kind: RepositoryKind) {
        this.composerData = Composer.createFromPath(path);
    }

    static create(path: string, kind: RepositoryKind) {
        return new Repository(path, kind);
    }

    get name() {
        return basename(this.path);
    }

    get composer() {
        return this.composerData;
    }

    get packages() {
        return this.composer.packages('all');
    }

    get files() {
        if (!this.loadedFiles) {
            this.loadFiles();
        }

        return this.fileList;
    }

    getFile(relativeName: string): RepositoryFile | null {
        return this.files.find(f => f.relativeName === relativeName) || null;
    }

    hasFile(file: RepositoryFile): boolean {
        return this.files.find(f => f.relativeName === file.relativeName) !== undefined;
    }

    hasIssues() {
        return this.issues.length > 0;
    }

    isSkeleton() {
        return this.kind === RepositoryKind.SKELETON;
    }

    isPackage() {
        return this.kind === RepositoryKind.PACKAGE;
    }

    protected getFileList(directory: string, basePath: string | null = null): RepositoryFile[] {
        const result: RepositoryFile[] = [];

        getFileList(directory, basePath, true)
            .filter(fn => !app.shouldIgnoreFile(fn.replace(`${basePath}/`, '')))
            .forEach(fqName => {
                const rfile = new RepositoryFile(this, fqName);

                const relativeName = basePath ? fqName.replace(`${basePath}/`, '') : fqName;
                const isFile = !isDirectory(fqName);

                if (!result.find(item => item.relativeName === relativeName)) {
                    result.push(
                        rfile.withOptions(
                            app.shouldIgnoreFile(relativeName),
                            rfile.isFile() && app.shouldCompareFile(fqName),
                            app.getFileScoreRequirements(fqName),
                        ),
                    );
                    //     {
                    //     isFile: !isPath,
                    //     name: fqName,
                    //     relativeName: relativeName,
                    //     filesize: fileSize(fqName),
                    //     shouldIgnore: app.shouldIgnoreFile(relativeName),
                    //     shouldCompare: !isPath && app.shouldCompareFile(fqName),
                    //     requiredScores: app.getFileScoreRequirements(fqName),
                    // });

                    if (rfile.isDirectory()) {
                        result.push(...this.getFileList(fqName, basePath ?? directory));
                    }
                }
            });

        return result;
    }

    public loadFiles() {
        this.fileList = this.getFileList(this.path, this.path);
        this.loadedFiles = true;

        return this;
    }
}
