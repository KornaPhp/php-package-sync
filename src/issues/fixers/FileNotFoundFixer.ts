/* eslint-disable no-unused-vars */

import { existsSync, writeFileSync } from 'fs';
import { basename } from 'path';
import { ComparisonKind } from '../../types/FileComparisonResult';
import { File } from '../../lib/File';
import { Fixer } from './Fixer';

export class FileNotFoundFixer extends Fixer {
    public static handles = [ComparisonKind.FILE_NOT_FOUND];

    public fix(): boolean {
        if (!this.shouldPerformFix()) {
            return false;
        }

        const relativeFn: string = this.issue.srcFile?.relativeName ?? this.issue.name;

        if (!existsSync(`${this.issue.repository.path}/${relativeFn}`)) {
            const data = File.read(`${this.issue.skeleton.path}/${relativeFn}`)
                .processTemplate(basename(this.issue.repository.path));

            writeFileSync(`${this.issue.repository.path}/${relativeFn}`, data, { encoding: 'utf-8' });
        }

        this.issue.resolve(this)
            .addResolvedNote(`copied file '${relativeFn}' from skeleton`);

        return true;
    }

    public static prettyName(): string {
        return 'create-file';
    }
}
