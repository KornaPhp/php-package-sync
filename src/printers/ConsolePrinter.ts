import { app } from '../Application';
import { Repository } from '../lib/Repository';

import Table from 'cli-table3';
import { ComparisonKind } from '../types/FileComparisonResult';

const chalk = require('chalk');

export class ConsolePrinter {
    public static kindColor(kind: ComparisonKind) {
        const colors = {
            [ComparisonKind.DIRECTORY_NOT_FOUND]: '#7DD3FC', //#3B82F6',
            [ComparisonKind.FILE_NOT_FOUND]: '#7DD3FC', //#3B82F6',
            [ComparisonKind.PACKAGE_NOT_USED]: '#818CF8',
            [ComparisonKind.PACKAGE_SCRIPT_NOT_FOUND]: '#10B981',
            [ComparisonKind.PACKAGE_VERSION_MISMATCH]: '#A78BFA',
            [ComparisonKind.FILE_NOT_SIMILAR_ENOUGH]: '#0EA5E9', //#475569',
            [ComparisonKind.ALLOWED_SIZE_DIFFERENCE_EXCEEDED]: '#64748B',
            [ComparisonKind.FILE_NOT_IN_SKELETON]: '#64748B',
        };

        const color = colors[kind] ?? '#FAFAF9';

        return chalk.hex(color);
    }

    public static printRepositoryIssues(repo: Repository) {
        const table = new Table({
            head: ['issue', 'score', 'filename', 'fixers', 'notes'],
            chars: { mid: '', 'left-mid': '', 'mid-mid': '', 'right-mid': '' },
            style: {
                head: [], //disable colors in header cells
                border: [], //disable colors for the border
            },
            colWidths: [15, 10, 50, 30, 30],
        });

        table.push(['-'.padEnd(12, '-'), '-'.padEnd(8, '-'), '-'.padEnd(45, '-'), '-'.padEnd(25, '-'), '-'.padEnd(25, '-')]);

        repo.issues
            .filter(issue => !issue.resolved)
            .filter(issue => !app.config.ignoreNames.includes(issue.name))
            .filter(issue => !app.config.skipComparisons.includes(issue.name))
            .filter(issue => !app.config.issues.ignored[issue.kind]?.includes(issue.name) ?? true)
            .sort((a, b) => (a.kind + a.score).localeCompare(b.kind + b.score))
            .forEach(issue => {
                table.push([
                    this.kindColor(issue.kind)(issue.kind),
                    issue.score,
                    this.kindColor(issue.kind)(issue.name),
                    issue.availableFixers.join(', '),
                    issue.note?.toString() ?? '',
                ]);
            });

        process.stdout.write(table.toString() + '\n');
    }

    public static printRepositoryFixerResults(repo: Repository) {
        const table = new Table({
            head: ['filename', 'fixer', 'status'],
            chars: { mid: '', 'left-mid': '', 'mid-mid': '', 'right-mid': '' },
            style: {
                head: [], //disable colors in header cells
                border: [], //disable colors for the border
            },
            colWidths: [40, 15, 65],
        });

        table.push(['-'.padEnd(35, '-'), '-'.padEnd(12, '-'), '-'.padEnd(55, '-')]);

        repo.issues
            //.filter(issue => issue.resolved)
            .filter(issue => !app.config.ignoreNames.includes(issue.name))
            .filter(issue => !app.config.skipComparisons.includes(issue.name))
            .filter(issue => !app.config.issues.ignored[issue.kind]?.includes(issue.name) ?? true)
            .sort((a, b) => a.kind.localeCompare(b.kind))
            .forEach(issue => {
                table.push([this.kindColor(issue.kind)(issue.name), issue.resolvedByFixer, issue.resolvedNotes.join('; ')]);
            });

        process.stdout.write(table.toString() + '\n');
    }
}
