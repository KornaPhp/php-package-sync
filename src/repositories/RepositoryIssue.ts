/* eslint-disable no-unused-vars */

import { ComparisonKind } from '../types/FileComparisonResult';
import { Repository } from './Repository';
import { RepositoryFile } from './RepositoryFile';
//import Fixer from './fixers/Fixer';
import { classOf } from '../lib/helpers';

export class RepositoryIssue {
    protected _fixers: any[] = [];
    public resolvedByFixer = 'none';
    public resolvedNotes: string[] = [];
    public pending = false;

    constructor(
        public result: any,
        public name: string,
        public srcFile: RepositoryFile | null,
        public destFile: RepositoryFile | null,
        public skeleton: Repository,
        public repository: Repository,
        public resolved: boolean = false,
        public note: string | null = null,
        public context: any | null = null,
    ) {
        //
    }

    get fixers() {
        const same = (a, b, method) => (a[method]() && b[method]()) || (!a[method]() && !b[method]());
        const isTrue = (a, method) => (a[method]() ? 1 : -1);
        const methodCompare = (a, b, method) => (same(a, b, method) ? 0 : isTrue(a, method));

        return this._fixers.sort((a, b) => methodCompare(a, b, 'runsFixers'))
            .sort((a, b) => methodCompare(a, b, 'isRisky'));
    }

    get kind(): ComparisonKind {
        return this.result.kind;
    }

    get score(): number | string {
        if (this.result.score === 0) {
            return '-';
        }

        if (typeof this.result.score === 'number') {
            return this.result.score.toFixed(3);
        }

        return this.result.score;
    }

    get sourcefile(): RepositoryFile {
        return <RepositoryFile>this.srcFile;
    }

    get targetfile(): RepositoryFile {
        return <RepositoryFile>this.destFile;
    }

    public addFixer(fixer: any) {
        if (this.fixers.find(f => f.getName() === fixer.getName()) === undefined) {
            this.fixers.push(fixer);
        }

        return this;
    }

    public is(kind: ComparisonKind): boolean {
        return this.kind === kind;
    }

    public resolve(resolvedByFixer: string | any) {
        if (typeof resolvedByFixer !== 'string') {
            resolvedByFixer = classOf(resolvedByFixer)
                .prettyName();
        }

        this.resolvedByFixer = <string>resolvedByFixer;
        this.resolved = true;

        return this;
    }

    public addResolvedNote(note: string, prepend = false) {
        if (prepend) {
            this.resolvedNotes.unshift(note);
            return this;
        }

        this.resolvedNotes.push(note);

        return this;
    }

    public disableFixers(names: string[] | null = null) {
        if (names === null) {
            names = this.fixers.map(fixer => fixer.getName());
        }

        this.fixers.forEach(fixer => {
            if (names?.includes(fixer.getName())) {
                fixer.disable();
            }
        });

        return this;
    }

    public enableFixers(names: string[] | null = null) {
        if (names === null) {
            names = this.fixers.map(fixer => fixer.getName());
        }

        this.fixers.forEach(fixer => {
            if (names?.includes(fixer.getName())) {
                fixer.enable();
            }
        });

        return this;
    }
}
