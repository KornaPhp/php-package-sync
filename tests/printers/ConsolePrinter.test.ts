/* eslint-disable no-undef */
/* eslint-disable @typescript-eslint/ban-ts-comment */
import { FileNotFoundFixer } from '../../src/fixers/FileNotFoundFixer';
import { ConsolePrinter } from '../../src/printers/ConsolePrinter';
import { Repository, RepositoryKind } from '../../src/repositories/Repository';
import { RepositoryIssue } from '../../src/repositories/RepositoryIssue';
import { ComparisonKind } from '../../src/types/FileComparisonResult';

const stripAnsi = require('strip-ansi');

let skeleton: Repository, repo: Repository, issue: RepositoryIssue, fixer: FileNotFoundFixer;

beforeEach(() => {
    skeleton = Repository.create(__dirname + '/../data/test-skeleton', RepositoryKind.SKELETON);
    repo = Repository.create(__dirname + '/../data/test-package-2', RepositoryKind.PACKAGE);
    //repo.loadFiles();

    issue = new RepositoryIssue({ kind: ComparisonKind.FILE_NOT_FOUND, score: 0 }, 'testfile1.txt', null, null, skeleton, repo);
    fixer = new FileNotFoundFixer(issue);

    issue.fixers.push(fixer);

    repo.issues.push(issue);
});

it('prints a summary of fixers', () => {
    const output = ConsolePrinter.printFixerSummary([fixer])
        .toString()
        .toLowerCase()
        .trim();

    expect(stripAnsi(output))
        .toMatchSnapshot();
});

it("prints a repository's issues", () => {
    const output = ConsolePrinter.printRepositoryIssues(repo)
        .toString()
        .toLowerCase()
        .trim();

    expect(stripAnsi(output))
        .toMatchSnapshot();
});

it("prints repository's fixed issues", () => {
    repo.issues[0].resolve(repo.issues[0].fixers[0]);

    const output = ConsolePrinter.printRepositoryFixerResults(repo)
        .toString()
        .toLowerCase()
        .trim();

    expect(stripAnsi(output))
        .toMatchSnapshot();
});
