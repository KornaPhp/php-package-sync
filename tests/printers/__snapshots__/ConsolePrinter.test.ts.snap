// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`prints a repository's issues 1`] = `
"┌───────────────┬────────┬─────────────────────────────────────────────┬──────────────────────────────┐
│ issue         │ score  │ filename                                    │ fixers                       │
│ ------------  │ -----  │ ------------------------------------------  │ ---------------------------  │
│ missing_file  │ -      │ testfile1.txt                               │ create-file                  │
└───────────────┴────────┴─────────────────────────────────────────────┴──────────────────────────────┘"
`;

exports[`prints a summary of fixers 1`] = `
"┌──────────────────┬────────┬───────────────────────────────────────────────────────────────────────────┐
│ fixer name       │ type   │ description                                                               │
│ ---------------  │ -----  │ ------------------------------------------------------------------------  │
│ create-file      │        │ copies a file from the skeleton repository into the package repository.   │
└──────────────────┴────────┴───────────────────────────────────────────────────────────────────────────┘"
`;

exports[`prints repository's fixed issues 1`] = `
"┌────────────────────────────────────────┬───────────────┬─────────────────────────────────────────────────────────────────┐
│ filename                               │ fixer         │ status                                                          │
│ -------------------------------------  │ ------------  │ --------------------------------------------------------------  │
│ testfile1.txt                          │ create-file   │                                                                 │
└────────────────────────────────────────┴───────────────┴─────────────────────────────────────────────────────────────────┘"
`;
